hbase-server/src/test/java/org/apache/hadoop/hbase/master/TestHMasterAsyncRegistryRPCs.java

```diff
@@ -0,0 +1,201 @@
+import org.apache.hadoop.hbase.shaded.protobuf.generated.MasterProtos.GetPortInfoResponse;
+import org.apache.hadoop.hbase.shaded.protobuf.generated.MasterProtos.GetMetaLocationsRequest;
+import org.apache.hadoop.hbase.shaded.protobuf.generated.MasterProtos.GetMetaLocationsResponse;
+import org.apache.hadoop.hbase.shaded.protobuf.generated.MasterProtos.IsActiveRequest;
+import org.apache.hadoop.hbase.shaded.protobuf.generated.MasterProtos.IsActiveResponse;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+@Category({ MasterTests.class})
+public class TestHMasterAsyncRegistryRPCs {
+    private static final Logger LOG = LoggerFactory.getLogger(TestHMasterAsyncRegistryRPCs.class);
+
+    private static final HBaseTestingUtility TEST_UTIL = new HBaseTestingUtility();
+    private static RpcClient rpcClient;
+    private static HMaster activeMaster;
+    private static List<HMaster> standByMasters;
+
+    @BeforeClass
+    public static void setUp() throws Exception {
+        Configuration conf = TEST_UTIL.getConfiguration();
+        conf.set(HConstants.MASTER_PORT, "0");
+        conf.setStrings(HConstants.ZOOKEEPER_ZNODE_PARENT, "/metacachetest");
+        TEST_UTIL.startMiniCluster(2);
+        activeMaster = TEST_UTIL.getHBaseCluster().getMaster();
+        standByMasters = new ArrayList<>();
+        // Create a few standby masters.
+        for (int i = 0; i < 2; i++) {
+            standByMasters.add(TEST_UTIL.getHBaseCluster().startMaster().getMaster());
+        }
+        rpcClient = RpcClientFactory.createClient(conf, HConstants.CLUSTER_ID_DEFAULT);
+    }
+
+    @AfterClass
+    public static void tearDown() throws Exception {
+        if (rpcClient != null) {
+            rpcClient.close();
+        }
+        TEST_UTIL.shutdownMiniCluster();
+    }
+
+    /**
+     * Verifies that getMetaReplicaIdFromPath() parses the full meta znode paths correctly.
+     * @throws IOException
+     */
+    @Test
+    public void TestGetMetaReplicaIdFromPath() throws IOException {
+        ZKWatcher zk = TEST_UTIL.getZooKeeperWatcher();
+        String metaZnode= ZNodePaths.joinZNode(zk.getZNodePaths().baseZNode, zk.getZNodePaths().metaZNodePrefix);
+        assertEquals(0, zk.getZNodePaths().getMetaReplicaIdFromPath(metaZnode));
+        for (int i = 1; i < 10; i++) {
+            assertEquals(i, zk.getZNodePaths().getMetaReplicaIdFromPath(metaZnode + "-" +i));
+        }
+        for (String suffix : new String[]{"foo", "1234", "foo-123", "123-foo-234"}) {
+            try {
+                final String input = metaZnode + suffix;
+                zk.getZNodePaths().getMetaReplicaIdFromZnode(input);
+                fail("Exception not hit getMetaReplicaIdFromZnode(): " +  input);
+            } catch (NumberFormatException e) {
+                // Expected
```

Comment:
nit: may be we can use assertEquals with NFE's error message if available?

Reply:
I think it's fine as it is, or if you prefer you can use mockito's expected functionality:

`@Test(expected = NumberFormatException.class)`